Sources/SwiftFusion/Optimizers/NLCG.swift

```diff
@@ -0,0 +1,94 @@
+// Copyright 2019 The SwiftFusion Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/// Non-Linear Conjugate Gradient (NLCG) optimizer.
+///
+/// An optimizer that implements NLCG second order optimizer
+
+public class NLCG<Model: Differentiable>
+  where Model.TangentVector: VectorProtocol & ElementaryFunctions & KeyPathIterable,
+  Model.TangentVector.VectorSpaceScalar == Double {
+  public typealias Model = Model
+  /// The set of steps taken.
+  public var step: Int = 0
+  public var precision: Double = 1e-10
+  public var max_iteration: Int = 400
+  
+  public init(
+    for _: __shared Model, precision p: Double = 1e-10, max_iteration maxiter: Int = 400) {
+    
+    precision = p
+    max_iteration = maxiter
+  }
+
+  func dot<T: Differentiable>(_ for: T, _ a: T.TangentVector, _ b: T.TangentVector) -> Double where T.TangentVector: KeyPathIterable {
+    a.recursivelyAllWritableKeyPaths(to: Double.self).map { a[keyPath: $0] * b[keyPath: $0] }.reduce(0.0, {$0 + $1})
+  }
+  
+  public func optimize(loss f: @differentiable @escaping (Model) -> Model.TangentVector.VectorSpaceScalar, model x_in: inout Model) {
+    step = 0
+    
+    let x_0 = x_in
+    
+    let dx_0 = gradient(at: x_0, in: f)
+    
+    let a_0 = 1.0
+    // a_0 = argmin(f(x_0+a*dx_0))
+    
+    var x_1 = x_0
+    x_1.move(along: dx_0.scaled(by: a_0))
+    
+    var x_n = x_1
+    var dx_n_1 = dx_0
+    var s = dx_0
+    
+    while step < max_iteration {
+      let dx = gradient(at: x_n, in: f)
+      
+      // Fletcher-Reeves
+      // TODO: `.dot` needs to be implemented by iterating over keyPath
+      let beta: Double = dot(x_in, dx, dx) / dot(x_in, dx_n_1, dx_n_1)
+      
+      // s_n = \delta x_n + \beta_n s_{n-1}
+      s = dx + s.scaled(by: beta)
+      
+//      debugPrint("dx = ", dx)
+//      debugPrint("s = ", s)
+//      debugPrint("x_n = ", x_n)
+      // Line search
+      // let a = argmin(f(x_n + a * s))
+      let f_a: @differentiable (_ a: Double) -> Model.TangentVector.VectorSpaceScalar = { a in
+        var x = x_n
+        let x_1 = x.withDerivative({ (dx: inout Model.TangentVector.TangentVector ) in
+          dx = s.scaled(by: a)
+          x.move(along: s.scaled(by: a))
+        })
+        return f(x_1)
+      }
```

Comment:
This function raises a compilation error:
```swift
error: function is not differentiable
      let f_a: @differentiable (_ a: Double) -> Model.TangentVector.VectorSpaceScalar = { a in
                                                                                        ^~~~~~
Sources/SwiftFusion/Optimizers/NLCG.swift:72:36: note: cannot differentiate through a non-differentiable argument; do you want to use 'withoutDerivative(at:)'?
        let x_1 = x.withDerivative({ (dx: inout Model.TangentVector.TangentVector ) in
```

Reply:
It seems not trivial to compute the derivative of arithmetic of tangent vectors? Here is what I want to do:
```swift
f_a = { a in x_n + a * s }
```

Reply:
I have a direct answer to your immediate question, but I also have a higher-level question about why you need to do this.

**My question**

It seems that you are trying to use gradient descent to do the line search. That doesn't seem right to me. Typically, line searches find the minimum by directly evaluating the function at a few points. If you do that, then you don't need to calculate this derivative.

**Answer to your question**

I believe the reason this is failing is that the vector space `.scaled` requirement (https://github.com/apple/swift/blob/0f3775689b546b03602007efb2be1dadafd001a8/stdlib/public/Differentiation/DifferentiationSupport.swift#L85) does not have `@differentiable`, so the compiler doesn't know what the derivative of `.scaled` is when you're operating over a generic vector space type.

This is something we could improve from the Swift AD side, maybe with a `DifferentiableVectorSpace` that has differentiable requirements.

In the meantime, you can work around it by calculating the gradient manually or by defining some custom differentiable add and scale functions.

Example how to calculate the gradient manually:

```swift
protocol InnerProductSpace: VectorProtocol {
  func dot(_ other: Self) -> VectorSpaceScalar
}

/// Returns the gradient of `f(x0 + s * a)` with respect to `a`.
func gradientAlongLine<T: Differentiable & InnerProductSpace>(
  _ f: @differentiable (T) -> T.VectorSpaceScalar,
  _ x0: T,
  _ s: T,
  _ a: T.VectorSpaceScalar
) -> T.VectorSpaceScalar
where
  T.TangentVector == T,
  T.VectorSpaceScalar == T.VectorSpaceScalar.TangentVector,
  T.VectorSpaceScalar: FloatingPoint
{
  let x = x0 + s.scaled(by: a)
  return gradient(at: x, in: f).dot(s)
}
```

Example how to define custom add and scaled derivatives:

```swift
public protocol InnerProductSpace: VectorProtocol {
  func dot(_ other: Self) -> VectorSpaceScalar
}

extension InnerProductSpace where
  Self: Differentiable,
  TangentVector == Self,
  VectorSpaceScalar: Differentiable,
  VectorSpaceScalar.TangentVector == VectorSpaceScalar
{
  @differentiable
  public func differentiableAdd(_ other: Self) -> Self {
    self + other
  }

  @derivative(of: differentiableAdd)
  @usableFromInline
  func vjpDifferentiableAdd(_ other: Self) -> (
    value: Self,
    pullback: (TangentVector) -> (TangentVector, TangentVector)
  ) {
    func pullback(_ v: TangentVector) -> (TangentVector, TangentVector) {
      (v, v)
    }
    return (differentiableAdd(other), pullback)
  }

  @differentiable
  public func differentiableScaled(by scalar: VectorSpaceScalar) -> Self {
    self.scaled(by: scalar)
  }

  @derivative(of: differentiableScaled)
  @usableFromInline
  func vjpDifferentiableScaled(by scalar: VectorSpaceScalar) -> (
    value: Self,
    pullback: (TangentVector) -> (TangentVector, VectorSpaceScalar)
  ) {
    func pullback(_ v: TangentVector) -> (TangentVector, VectorSpaceScalar) {
      (v.scaled(by: scalar), v.dot(self))
    }
    return (differentiableScaled(by: scalar), pullback)
  }
}

/// Returns the gradient of `f(x0 + s * a)` with respect to `a`.
func gradientAlongLine<T: Differentiable & InnerProductSpace>(
  _ f: @differentiable (T) -> T.VectorSpaceScalar,
  _ x0: T,
  _ s: T,
  _ a: T.VectorSpaceScalar
) -> T.VectorSpaceScalar
where
  T.TangentVector == T,
  T.VectorSpaceScalar == T.VectorSpaceScalar.TangentVector,
  T.VectorSpaceScalar: FloatingPoint
{
  gradient(at: a) { a in
    f(x0.differentiableAdd(s.differentiableScaled(by: a)))
  }
}
```

Reply:
I think (Fan,  please confirm) the line search in CGLS is in a particular direction (directional gradient), so the 1D "line search" needs to have that direction.

Reply:
@dellaert I think you mean NLCG? For NLCG, yes, the line search is an 1D argmax in a particular direction.

Reply:
@marcrasi Yes you are right that it is normally done with evaluation on some points :) In GTSAM it is done with Golden Section Search. This problem is more like what I encountered in a conceptual evaluation of the argmax operation in a test-driven manner.